README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# rwlts <img src="inst/extdata/img/logo.png" align="right" width="120"/>

R Client Library for Web Land Trajectory Service (WLTS)

<!-- badges: start -->

[![Software
License](https://img.shields.io/badge/license-MIT-green)](https://github.com/brazil-data-cube/rstac/blob/master/LICENSE)
[![Build Status](https://drone.dpi.inpe.br/api/badges/brazil-data-cube/rwlts/status.svg)](https://drone.dpi.inpe.br/brazil-data-cube/rwlts)
[![codecov](https://codecov.io/gh/brazil-data-cube/rwlts/branch/main/graph/badge.svg?token=6WTJI6K93Y)](https://codecov.io/gh/brazil-data-cube/rwlts)
[![Software Life
Cycle](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html)
[![Join us at
Discord](https://img.shields.io/discord/689541907621085198?logo=discord&logoColor=ffffff&color=7389D8)](https://discord.com/channels/689541907621085198#)

<!-- badges: end -->

# About

Information on Land Use and Land Cover (LULC) is essential to support
governments in making decisions about the impact of human activities on
the environment, planning the use of natural resources, conserving
biodiversity and monitoring climate change.

Currently, several projects systematically provide information on the
dynamics of land use and cover. Well known projects include PRODES,
DETER and TerraClass. These projects are developed by INPE and they
produce information on land use and coverage used by the Brazilian
Government to make public policy decisions. Besides these projects there
are other initiatives from universities and space agencies devoted to
the creation of national and global maps.

Although these projects follow open data policies and provide a rich
collection of data, there is still a gap in tools that facilitate the
integrated use of these collections. Each project adopts its own land
use and land cover classification system, providing different class
names and meanings for the elements of these collections. The forms of
distribution of project data can be carried out in different ways,
through files or web services. In addition, the data has different
spatial and temporal resolutions and storage systems (raster or vector).

In this context, the **W**eb **L**and **T**rajectory **S**ervice (WLTS)
is a service that aims to facilitate the access to these approach
consists of using a data model that defines a minimum set of temporal
and spatial information to represent different sources and types of
data, but with a focus on land use and land cover.

WLTS can be used in a variety of applications, such as validating land
cover data sets, selecting training samples to support Machine Learning
algorithms used in the generation of new classification maps.

If you want to know more about WLTS service, please, take a look at its
[specification](https://github.com/brazil-data-cube/wlts-spec).

## Installation

To install the development version of `rwlts`, run the following
commands:

``` r
# load necessary libraries
library(devtools)
devtools::install_github("brazil-data-cube/rwlts")
```

Importing `rwlts` package:

``` r
library(rwlts)
```

## Usage

`rwlts` implements the following WLTS operations:

| **WLTS** operations     | `rwlts` functions                          |
|:------------------------|:-------------------------------------------|
| `/list_collections`     | `list_collections(URL,)`                   |
| `/describe_collections` | `describe_collection(URL, collection_id)`  |
| `/trajectory`           | `get_trajectory(URL, latitude, longitude)` |

These functions can be used to retrieve information from a WLTS API
service. The code bellow creates a `wlts` object and list the available
collections of the WLTS API of the Brazil Data Cube project of the
Brazilian National Institute for Space Research INPE.

### List Collections

The first operation, `list_collections`, retrieves all available
collections in WLTS service.

``` r
wlts_bdc <- "https://brazildatacube.dpi.inpe.br/wlts/"

list_collections(wlts_bdc)
#>  [1] "lapig_areas_pastagem"        "terraclass_amazonia"        
#>  [3] "deter_amazonia_legal"        "deter_cerrado"              
#>  [5] "ibge_cobertura_uso_terra"    "prodes_amazonia_legal"      
#>  [7] "prodes_cerrado"              "mapbiomas_caatinga-v5"      
#>  [9] "mapbiomas_cerrado-v5"        "mapbiomas_amazonia-v5"      
#> [11] "terraclass_amazonia-v2"      "mapbiomas_pantanal-v5"      
#> [13] "mapbiomas_pampa-v5"          "terraclass_cerrado"         
#> [15] "mapbiomas_mata_atlantica-v5"
```

### Describe Collection

Each collection returned by the WLTS service can be used for retrieving
LULC trajectories. To get information about the collections, we use the
`describe_collection` function. This function returns the metadata of a
given collection, which includes the description and specific details.
The metadata also describes the spatio-temporal extent of the
collection.

In the code below, we retrieve the metadata from the
`deter_amazonia_legal` collection using the `describe_collection`
function.

``` r
describe_collection(wlts_bdc, "deter_amazonia_legal")
#> $classification_system
#> $classification_system$classification_system_id
#> [1] "21"
#> 
#> $classification_system$classification_system_name
#> [1] "DETER Amazônia Legal"
#> 
#> $classification_system$classification_system_version
#> [1] "1.0"
#> 
#> $classification_system$type
#> [1] "Self"
#> 
#> 
#> $collection_type
#> [1] "Feature"
#> 
#> $description
#> [1] "Alertas de Desmatamento da Amazônia Legal."
#> 
#> $detail
#> [1] "O DETER é um levantamento rápido de alertas de evidências de alteração da cobertura florestal na Amazônia, feito pelo INPE. O DETER foi desenvolvido como um sistema de alerta para dar suporte à fiscalização e controle de desmatamento e da degradação florestal realizadas pelo Instituto Brasileiro do Meio Ambiente e dos Recursos Naturais Renováveis (IBAMA) e demais órgãos ligados a esta temática. Mais informações acesse: http://www.obt.inpe.br/OBT/assuntos/programas/amazonia/deter"
#> 
#> $name
#> [1] "deter_amazonia_legal"
#> 
#> $period
#> $period$end_date
#> [1] "2020-06-18"
#> 
#> $period$start_date
#> [1] "2016-08-02"
#> 
#> 
#> $resolution_unit
#> $resolution_unit$unit
#> [1] "DAY"
#> 
#> $resolution_unit$value
#> [1] 1
#> 
#> 
#> $spatial_extent
#> $spatial_extent$xmax
#> [1] -44.00039
#> 
#> $spatial_extent$xmin
#> [1] -73.54909
#> 
#> $spatial_extent$ymax
#> [1] 4.555376
#> 
#> $spatial_extent$ymin
#> [1] -18.03644
```

### Trajectory

LULC trajectories can be extracted from collections. These LULC
trajectories in WLTS services are associated with a point (lat, long) in
geographic space. In `rwlts`, we can use the `get_trajectory` function
to retrieve trajectories. For example, in the code below, the point
`(-54, -12)` trajectory is retrieved from the `mapbiomas_amazonia-v5`
collection.

``` r
get_trajectory(wlts_bdc, 
               latitude = -12, 
               longitude = -54, 
               collections = "mapbiomas_amazonia-v5",
               config = httr::add_headers("x-api-key" = "change-me"))
#> $query
#> NULL
#> 
#> $result
#> # A tibble: 35 × 4
#>    class              collection            date  point_id
#>    <chr>              <chr>                 <chr>    <int>
#>  1 Formação Florestal mapbiomas_amazonia-v5 1985         1
#>  2 Formação Florestal mapbiomas_amazonia-v5 1986         1
#>  3 Formação Florestal mapbiomas_amazonia-v5 1987         1
#>  4 Formação Florestal mapbiomas_amazonia-v5 1988         1
#>  5 Formação Florestal mapbiomas_amazonia-v5 1989         1
#>  6 Formação Florestal mapbiomas_amazonia-v5 1990         1
#>  7 Formação Florestal mapbiomas_amazonia-v5 1991         1
#>  8 Formação Florestal mapbiomas_amazonia-v5 1992         1
#>  9 Formação Florestal mapbiomas_amazonia-v5 1993         1
#> 10 Formação Florestal mapbiomas_amazonia-v5 1994         1
#> # … with 25 more rows
#> 
#> attr(,"class")
#> [1] "wlts"
```

The `get_trajectory` function returns a `list` of class `wlts`. This
object contains the `query` and `result` attributes. The `query`
attribute stores the query performed to retrieve the data. By default,
it will be `NULL`. For this information to be stored, the
`query_info = TRUE` parameter is required. For example:

``` r
get_trajectory(wlts_bdc, 
               latitude = -12, 
               longitude = -54, 
               collections = "mapbiomas_amazonia-v5",
               query_info = TRUE,
               config = httr::add_headers("x-api-key" = "change-me"))
```

The `result` attribute stores the retrieved trajectories. The data is
stored in a `tibble` for easy manipulation, which has the columns:

-   `class`: LULC class;
-   `collection`: Data Collection;
-   `date`: Time instants of the trajectory;
-   `point_id`: ID of the point that was queried.

The `point_id` column of the result `tibble` is used in `rwlts` to
identify the entry point. This ID is necessary since the
`get_trajectory` function can be used with vectors as input. For
example, the code below retrieves the trajectory of data from the
`mapbiomas_amazonia-v5` collection for two points `(-54, -12)` and
`(-54, -11.01)`.

``` r
get_trajectory(wlts_bdc, 
               latitude = c(-12, -11.01), 
               longitude = c(-54, -54), 
               collections = "mapbiomas_amazonia-v5",
               config = httr::add_headers("x-api-key" = "change-me"))
#> $query
#> NULL
#> 
#> $result
#> # A tibble: 70 × 4
#>    class              collection            date  point_id
#>    <chr>              <chr>                 <chr>    <int>
#>  1 Formação Florestal mapbiomas_amazonia-v5 1985         1
#>  2 Formação Florestal mapbiomas_amazonia-v5 1986         1
#>  3 Formação Florestal mapbiomas_amazonia-v5 1987         1
#>  4 Formação Florestal mapbiomas_amazonia-v5 1988         1
#>  5 Formação Florestal mapbiomas_amazonia-v5 1989         1
#>  6 Formação Florestal mapbiomas_amazonia-v5 1990         1
#>  7 Formação Florestal mapbiomas_amazonia-v5 1991         1
#>  8 Formação Florestal mapbiomas_amazonia-v5 1992         1
#>  9 Formação Florestal mapbiomas_amazonia-v5 1993         1
#> 10 Formação Florestal mapbiomas_amazonia-v5 1994         1
#> # … with 60 more rows
#> 
#> attr(,"class")
#> [1] "wlts"
```

In this case, the point `(-54, -12)` will have `point_id` equal to 1,
while `(-54, -11.01)` will have `point_id` equal to 2.

In addition to multiple point retrieval, the `get_trajectory` function
allows multiple collections to be queried for the composition of the
trajectory. To do this, in the `collections` parameter, the collections
must be assigned. For example, data from the collections
`mapbiomas_amazonia-v5` and `terraclass_amazonia-v2` are retrieved in
the code below.

``` r
get_trajectory(wlts_bdc, 
               latitude = c(-12, -11.01), 
               longitude = c(-54, -54), 
               collections = c("mapbiomas_amazonia-v5", "terraclass_amazonia-v2"),
               config = httr::add_headers("x-api-key" = "change-me"))
#> $query
#> NULL
#> 
#> $result
#> # A tibble: 80 × 4
#>    class              collection            date  point_id
#>    <chr>              <chr>                 <chr>    <int>
#>  1 Formação Florestal mapbiomas_amazonia-v5 1985         1
#>  2 Formação Florestal mapbiomas_amazonia-v5 1986         1
#>  3 Formação Florestal mapbiomas_amazonia-v5 1987         1
#>  4 Formação Florestal mapbiomas_amazonia-v5 1988         1
#>  5 Formação Florestal mapbiomas_amazonia-v5 1989         1
#>  6 Formação Florestal mapbiomas_amazonia-v5 1990         1
#>  7 Formação Florestal mapbiomas_amazonia-v5 1991         1
#>  8 Formação Florestal mapbiomas_amazonia-v5 1992         1
#>  9 Formação Florestal mapbiomas_amazonia-v5 1993         1
#> 10 Formação Florestal mapbiomas_amazonia-v5 1994         1
#> # … with 70 more rows
#> 
#> attr(,"class")
#> [1] "wlts"
```

Finally, the `get_trajectory` function, through the `start_date` and
`end_date` parameters, allows you to specify the time intervals used in
the trajectory. To exemplify its use, in the code below, trajectories
are retrieved for the points `(-54, -12)` and `(-54, -11.01)`, from the
`mapbiomas_amazonia-v5` and `terraclass_amazonia-v2` collections in the
time interval `[2003-01-01, 2004-01-01]`.

``` r
get_trajectory(wlts_bdc, 
               latitude = c(-12, -11.01), 
               longitude = c(-54, -54), 
               start_date = "2003-01-01",
               end_date = "2004-01-01",
               collections = c("mapbiomas_amazonia-v5","terraclass_amazonia-v2"),
               config = httr::add_headers("x-api-key" = "change-me"))
#> $query
#> NULL
#> 
#> $result
#> # A tibble: 6 × 4
#>   class                                collection             date  point_id
#>   <chr>                                <chr>                  <chr>    <int>
#> 1 Formação Florestal                   mapbiomas_amazonia-v5  2003         1
#> 2 Formação Florestal                   mapbiomas_amazonia-v5  2004         1
#> 3 Vegetação Natural Florestal Primária terraclass_amazonia-v2 2004         1
#> 4 Formação Florestal                   mapbiomas_amazonia-v5  2003         2
#> 5 Formação Florestal                   mapbiomas_amazonia-v5  2004         2
#> 6 Vegetação Natural Florestal Primária terraclass_amazonia-v2 2004         2
#> 
#> attr(,"class")
#> [1] "wlts"
```

### Alluvial plot

To visualize the trajectories and fully understand their time dynamics,
the `rwlts` package implements the Alluvial-based visualization method.
To create this plot, use the `plot` function, as shown in the example
below:

``` r
# import data from package
data("mt_500_mapbiomas_cerrado")

plot(mt_500_mapbiomas_cerrado)
```

<img src="inst/extdata/img/README-unnamed-chunk-12-1.png" width="100%" />

Besides, you can fully customize the returned plot. This customization
is possible since the `plot` function returns a ggplot2 object.

``` r
library(ggplot2)
library(ggalluvial) # use to create plot
library(cowplot)    # use different theme

plot(mt_500_mapbiomas_cerrado, show_count = TRUE) + 
  cowplot::theme_minimal_hgrid() + 
  labs(title = "Changes in Primavera do Leste (2004-2014)",
       x = "Timeline",
       y = "Number of points",
       fill = "Class") + 
  theme(legend.position = "bottom",
        plot.title = element_text(hjust = 0.5)) + 
  scale_fill_manual(values = c("#129912",
                               "#32CD32",
                               "#BDB76B",
                               "#FFD966",
                               "#FFFFB2"),
                    labels = c("Formação Florestal",
                               "Formação Savânica",
                               "Outras Lavouras Temporárias",
                               "Pastagem",
                               "Soja"))
```

<img src="inst/extdata/img/README-unnamed-chunk-13-1.png" width="100%" />

> The numbers inside each bar correspond to the quantity of points
> extracted in each year. You can see that according to the change of
> LULC classes, the quantity of points in each class also changes.

# License

Copyright (C) 2021 INPE.

R client for WLTS is free software; you can redistribute it and/or
modify it under the terms of the MIT License; see LICENSE file for more
details.


NEWS.md
# Version 0.8.0 (2021-10-04)

- Change `sankey` to `alluvial`
- Collection can be provide as character vector
- Implement plot using s3 class
- Change version

# Version 0.6.0 (2021-09-03)

- Add Sankey Plot
- Add Drone integration
- Update documentation

# Version 0.4.0-0 (2020-12-14)

- Base documentation;
- Add CI support (unit tests and code coverage);
- Add support for all WLTS operations:

    - `/list_collections`
    - `/describe_collection`
    - `/trajectory`

- License: [MIT](https://raw.githubusercontent.com/brazil-data-cube/rwlts/master/LICENSE)


tests/testthat.R
library(testthat)
library(rwlts)

test_check("rwlts")


tests/testthat/helper-rwlts.R
library(vcr)
invisible(vcr::vcr_configure(
  dir    = "../fixtures",
  record = "new_episodes",
  write_disk_path = "."
))
vcr::check_cassette_names()


tests/testthat/test-examples.R
context("test_examples")

testthat::test_that("examples file", {
  vcr::use_cassette("examples_file", {

    # skip cran check test
    testthat::skip_on_cran()




  })
})


R/plot.R
#' @title Plots for rwlts package
#'
#' @name plot.wlts
#'
#' @description A set of visualization methods for trajectories extracted in
#'  rwlts package.
#'
#' @param x    a \code{tibble} object from class \code{wlts}.
#' @param type a \code{character} with the type of plot.
#'  Nowadays, only the "alluvial" plot is supported. By default is "alluvial".
#' @param ...         additional functions
#'
#' @examples
#' \dontrun{
#'  wlts_bdc <- "https://brazildatacube.dpi.inpe.br/wlts/"
#'
#'  wlts_tibble <- get_trajectory(
#'                      URL        = wlts_bdc,
#'                      latitude    = c(-12, -11),
#'                      longitude   = c(-54, -55),
#'                      collections = "mapbiomas_amazonia-v5",
#'                      start_date  = "2015-07-01",
#'                      end_date    = "2017-07-01",
#'                      config = httr::add_headers("x-api-key" = "BDC-KEY"))
#'
#'  plot(wlts_tibble)
#' }
#'
#' @return a \code{gg} object from ggplot2 package.
#'
#' @export
plot.wlts <- function(x, ..., type = "alluvial") {

  if (is.null(x$result))
    stop(paste("The result provided is empty, please check your query."),
         .call = FALSE)

  if (!requireNamespace("ggplot2", quietly = TRUE)) {
    stop(paste("ggplot2 required for this function to work.",
               "Please install it."), call. = FALSE)
  }

  # dispatch for S3 methods
  class(x) <- c(type, class(x))
  plot(x, ...)
}

#' @rdname plot.wlts
#' @param show_count  a \code{logical} parameter, if true, is added
#' the number of points on each bar. The default value is FALSE.
#' @export
plot.alluvial <- function(x, ..., show_count = FALSE) {

  if (!requireNamespace("ggalluvial", quietly = TRUE)) {
    stop(paste("ggalluvial required for this function to work.",
               "Please install it."), call. = FALSE)
  }

  traj_data_sankey <- x$result %>%
    dplyr::mutate(date = as.factor(as.numeric(date)),
                  class = as.factor(class)) %>%
    dplyr::arrange(date) %>%
    dplyr::group_by(class, date) %>%
    dplyr::mutate(n_points = dplyr::n()) %>%
    dplyr::ungroup()

  g <- ggplot2::ggplot(traj_data_sankey,
                       ggplot2::aes_string(x = "date",
                                           stratum = "class",
                                           alluvium = "point_id",
                                           fill = "class",
                                           label = "class")) +
    ggalluvial::geom_flow(stat = "alluvium",
                          lode.guidance = "frontback",
                          curve_type = "xspline") +
    ggalluvial::geom_stratum() +
    ggplot2::scale_y_continuous(expand = c(0, 0)) +
    ggplot2::theme(legend.position = "bottom")

  if (show_count)
    g <- g + ggplot2::geom_text(stat = "stratum",
                                ggplot2::aes_string(label = "n_points"))

  if (length(unique(x$result$collection)) > 1)
    g <- g + ggplot2::facet_wrap(~ traj_data_sankey$collection)

  return(g)
}


R/utils.R
#' @title Validates the location of geographic coordinates
#' @name .check_location
#'
#' @description Checks if the values defined for the location are valid
#' geographical coordinates and are within the range of values
#'
#' @param longitude \code{numeric} Longitude in WGS84 coordinate system.
#' @param latitude  \code{numeric} Latitude in WGS84 coordinate system.
.check_location <- function(latitude, longitude) {

  if (length(latitude) != length(longitude))
    stop(paste("The number of points in latitude and",
               "longitude should be the same length."), call. = FALSE)

  if (any(latitude < -90 | latitude > 90))
    stop("latitude is out of range (-90, 90)", call. = FALSE)

  if (any(longitude < -180 | longitude > 180))
    stop("longitude is out of range (-180, 180)", call. = FALSE)
}

#' @title Validates the dates
#' @name .check_datetime
#'
#' @description Checks if the dates are in RFC 3339 format and if they are in a
#'  valid start and end date relationship
#'
#' @param start_date \code{character} Start date in RFC 3339 format
#' @param end_date   \code{character} End date in RFC 3339 format
.check_datetime <- function(start_date, end_date) {

  pattern_rfc  <- "^\\d{4}-\\d{2}-\\d{2}?"
  check_status <- vapply(c(start_date, end_date), grepl,
                         pattern   = pattern_rfc,
                         perl      = TRUE,
                         FUN.VALUE = logical(1),
                         USE.NAMES = FALSE)

  if (!all(check_status))
    stop("The dates must be in the format of RFC 3339 (YYYY-MM-DD)")

  if (all(!is.null(start_date) & !is.null(end_date))) {
    if (start_date >= end_date)
      stop("The 'start_date' should be less than 'end_date'")
  }
}

#' @title Validates the dates relationship
#' @name .drop_na
#'
#' @description Check if they are in a valid start and end date relationship
#'
#' @param elements \code{character} A vector of character elements
#'
#' @return a filtered \code{list}.
.drop_na <- function(elements) {
  elements[-which(sapply(elements, is.null))]
}


R/request.R
#' @title Make an HTTP request
#'
#' @name request
#'
#' @description makes an HTTP request along with HTTP query parameters
#' for a web service.
#'
#' @param URL \code{character} URL of the WLTS Service.
#' @param ... \code{character} Parameters to httr::GET function.
#' @param query \code{list} A named list with HTTP query parameters.
#'
#' @return a \code{object} response from httr package.
request <- function(URL, ..., query = NULL) {

  tryCatch({
    response_obj <- httr::GET(URL, query = query, ...)
  },
  error = function(e) {
    stop("Error in requisition: ", e)
  })

  httr::content(response_obj)
}


#' @title Create a valid URL
#'
#' @name .build_url
#'
#' @description Checks if the URL has patterns that may cause query problems.
#'
#' @param URL        a \code{character} URL of the WLTS Service.
#' @param path       a \code{character} with URL path.
#' @param query      a \code{list} A named list with HTTP query parameters.
#' @param names_list a \code{character} vector with the names of query
#'  attributes.
#'
#' @return a \code{list} with \code{url} and \code{query} attributes.
.build_url <- function(URL, path, query = NULL, names_list = NULL) {
  url <- paste0(URL, path)
  url <- gsub("/$", "", url)

  names(query) <- names_list
  return(list(url = url, query = query))
}


R/wltr-package.R

#' @title wltr
#' @description An R client to the web land trajectory service (WLTS)
#'
#' @section WLTS API:
#'
#' Implements an R interface to a web land trajectory service (WLTS)
#' that offers a land trajectory time series of remote sensing samples API.
#'
#' The WLTS API has three commands:
#' \itemize{
#'    \item `list_collections`: lists all available collections on WLTS Service
#'    \item `trajectory`: Returns the Land Use and Land Cover for a spatio-temporal location
#'    \item `describe_collection`: retrieves the metadata of a collection
#' }
#'
#' @docType package
#' @name wlts-package
#' @aliases wlts
"_PACKAGE"
NULL

#' Pipe
#'
#' Magrittr compound assignment pipe-operator.
#'
#' @importFrom magrittr %>%
#' @name %>%
#' @rdname pipe
#' @param lhs,rhs A visualization and a function to apply to it.
#' @export
NULL

#' @importFrom dplyr mutate arrange ungroup group_by
#' @importFrom ggplot2 ggplot aes_string scale_y_continuous theme facet_wrap
#' geom_text
#' @importFrom ggalluvial geom_flow geom_stratum
NULL


R/rwlts-package.R
#' @title wltr
#' @description An R client to the web land trajectory service (WLTS)
#'
#' @section WLTS API:
#'
#' Implements an R interface to a web land trajectory service (WLTS)
#' that offers a land trajectory time series of remote sensing samples API.
#'
#' The WLTS API has three commands:
#' \itemize{
#'    \item `list_collections`: lists all available collections on WLTS Service
#'    \item `trajectory`: Returns the Land Use and Land Cover for a spatio-temporal location
#'    \item `describe_collection`: retrieves the metadata of a collection
#' }
#'
#' @docType package
#' @name wlts-package
#' @aliases wlts
"_PACKAGE"

#' Pipe
#'
#' Magrittr compound assignment pipe-operator.
#'
#' @importFrom magrittr %>%
#' @name %>%
#' @rdname pipe
#' @param lhs,rhs A visualization and a function to apply to it.
#' @export
NULL

#' @importFrom dplyr mutate arrange ungroup group_by
NULL


R/wlts.R
#' @title List Collections
#'
#' @description Retrieves the list of available data collections.
#'
#' @param URL         a \code{character} URL of the WLTS Service.
#' @param ...         a \code{list} Parameters to httr::GET function
#'
#' @examples
#' \dontrun{
#'  wlts_bdc <- "https://brazildatacube.dpi.inpe.br/wlts/"
#'
#'  list_collections(wlts_bdc)
#' }
#'
#' @return a \code{character} vector with the available data collections.
#'
#' @export
list_collections <- function(URL, ...) {

  if (is.null(URL))
    stop("The WLTS URL service must be provided.")

  # is this best way?
  url_obj <- .build_url(URL, path = "/list_collections")

  content <- request(url_obj$url, ...)
  unlist(content, use.names = FALSE)
}

#' @title Describe Collection
#'
#' @description Retrieves the metadata of a given data collection.
#'
#' @param URL           a \code{character} URL of the WLTS Service.
#' @param collection_id a \code{character} with identifier (name) of a
#'  collection.
#' @param ...           a \code{list} Parameters to httr::GET function.
#'
#' @examples
#' \dontrun{
#'  wlts_bdc <- "https://brazildatacube.dpi.inpe.br/wlts/"
#'
#'  describe_collection(wlts_bdc, "deter_amazonia_legal")
#' }
#'
#' @return a named \code{list} with the metadata of data collection.
#'
#' @export
describe_collection <- function(URL, collection_id, ...) {

  if (is.null(URL))
    stop("WLTS URL service must be provided.")

  if (is.null(collection_id))
    stop("collection_id must be provided.")

  url_obj <- .build_url(URL, path = "/describe_collection",
                        query = list(collection_id),
                        names_list = "collection_id")

  content <- request(url_obj$url, query = url_obj$query, ...)
  content
}

#' @title Get Trajectory
#'
#' @description Retrieves the land use and cover trajectories from the data
#'  collections given a location in space. The property \code{result} contains
#'  the feature identifier information, class, time, and the collection
#'  associated to the data item.
#'
#' @param URL         a \code{character} URL of the WLTS Service.
#' @param latitude    a \code{numeric} vector corresponding to latitude in
#'  WGS84 coordinate system.
#' @param longitude   a \code{numeric} vector corresponding to longitude in
#'  WGS84 coordinate system.
#' @param collections a \code{character} vector of identifier (name) of
#'  collections delimited by comma.
#' @param start_date  a \code{character} with the start date to be search.
#' @param end_date    a \code{character} with the end date to be search.
#' @param ...         a \code{list} Parameters to httr::GET function.
#' @param query_info  a \code{bolean} flag, if true query information is
#'  returned.
#'
#' @examples
#' \dontrun{
#'  wlts_bdc <- "https://brazildatacube.dpi.inpe.br/wlts/"
#'
#'  get_trajectory(wlts_bdc,
#'                 latitude = c(-12, -11.01),
#'                 longitude = c(-54, -54),
#'                 collections = "mapbiomas5_amazonia")
#'
#'  # date interval
#'  get_trajectory(URL        = wlts_bdc,
#'                latitude    = c(-12, -11),
#'                longitude   = c(-54, -55),
#'                collections = "mapbiomas5_amazonia",
#'                start_date  = "2015-07-01",
#'                end_date    = "2017-07-01")
#' }
#'
#' @return a \code{object} of wlts class with query (if \code{query_info} is
#'  provided) and a tibble with trajectory requested.
#'
#' @export
get_trajectory <- function(URL,
                           latitude,
                           longitude,
                           collections = NULL,
                           start_date  = NULL,
                           end_date    = NULL,
                           ...,
                           query_info = FALSE) {

  if (is.null(URL))
    stop("WLTS URL service must be provided.")

  # check if the latitude and longitude
  .check_location(latitude, longitude)

  if (any(!is.null(start_date) | !is.null(end_date)))
    .check_datetime(start_date, end_date)

  if (length(collections) > 1)
    collections <- paste(collections, collapse = ",")

  # build final url
  url_obj <- .build_url(URL, path  = "/trajectory",
                        query      = list(start_date, end_date, collections),
                        names_list = c("start_date", "end_date", "collections"))

  # create a list of content
  list_content <- lapply(seq_along(latitude), function(i) {
    url_obj$query[c("latitude", "longitude")] <- c(latitude[i], longitude[i])

    # TODO: adjust parent.frame
    content <- request(url_obj$url, query = url_obj$query, ...)
    content$result <- .build_wlts_tb(content$result, index = i)

    content
  })

  result_query <- NULL
  if (query_info)
    result_query <- sapply(list_content, `[[`, "query", simplify = FALSE)

  wlts_tb <- do.call(rbind, sapply(list_content, `[[`, "result",
                                   simplify = FALSE))

  structure(list(
    query  = result_query,
    result = wlts_tb),
    class  = "wlts")
}
